ElasticSearch customization and new property (#165)
mramendi/kie-docs: doc-content/jbpm-docs/src/main/asciidoc/ElasticSearch/integration-elasticsearch-customize-proc.adoc

[id='integration-elasticsearch-customize-proc_{context}']
= Customizing data for ElasticSearch

You can develop transformer classes to customize the data that {PRODUCT} sends to ElasticSearch.

Information about processes, cases, tasks, and task operations is available as _views_. {PRODUCT} includes the following view types:

* `CaseInstanceView`
* `ProcessInstanceView`
* `TaskInstanceView`
* `TaskOperationView`

You can see the definitions of these views in the https://github.com/kiegroup/jbpm/tree/{COMMUNITY_VERSION_FINAL}/jbpm-persistence/jbpm-persistence-api/src/main/java/org/jbpm/persistence/api/integration/model[GitHub repository].

Each view has a `getCompositeId()` method that returns an identifier. This identifier denotes a particular instance of a case, process, task, or task operation.

> Each time a process, case, task, or task operation is created, updated, or deleted, the {PROCESS_ENGINE} calls a _transformer_ and supplies the relevant view.

Reviewer: ```suggestion
Each time you create, update or delete a process, case, task, or task operation, the {PROCESS_ENGINE} calls a _transformer_ and supplies the relevant view.
```
Remove passive voice.

> The transformer must generate a `ESRequest` object. In the parameters of the constructor of this object, the transformer must supply the necessary information for the ElasticSearch request, including the index.

Author: I don't really like so many 's :) FIxing the article though.

> The definitions of the transformer classes and of the `ESRequest` class are available in the https://github.com/kiegroup/jbpm/tree/{COMMUNITY_VERSION_FINAL}/jbpm-event-emitters/jbpm-event-emitters-elasticsearch/src/main/java/org/jbpm/event/emitters/elasticsearch[GitHub repository].

Reviewer: ```suggestion
The definitions of the transformer classes and the `ESRequest` class are available in the https://github.com/kiegroup/jbpm/tree/{COMMUNITY_VERSION_FINAL}/jbpm-event-emitters/jbpm-event-emitters-elasticsearch/src/main/java/org/jbpm/event/emitters/elasticsearch[GitHub repository].
```

To create and use custom transformers, complete the following procedure.

.Procedure

. Create the Java source code for the following classes:

> ** `ESInstanceViewTransformer`: The transformer class. It provides `index()` and `update()` methods. Both of the methods take a view as a parameter and return a `ESRequest` object.

Reviewer: ```suggestion
** `ESInstanceViewTransformer`: The transformer class. It provides `index()` and `update()` methods. Both of the methods take a view as a parameter and return an `ESRequest` object.
```

+

> When a process, case, task, or task operation instance is first created, the {PROCESS_ENGINE} calls the `index()` method. For subsequent changes related to the same instance, the {PROCESS_ENGINE} calls the `update()` method.

Reviewer: ```suggestion
When you first create a process, case, task, or task operation, the {PROCESS_ENGINE} calls the `index()` method. For subsequent changes related to the same instance, the {PROCESS_ENGINE} calls the `update()` method.
```
Passive voice.

+
You can create different `ESInstanceViewTransformer` implementations for different view types.
+
** `ESInstanceViewTransformerFactory`: The transformer factory class. It returns an instance of the `ESInstanceViewTransformer` class for every view type.
+

> . In {CENTRAL}, enter your project and select the *Settings* -> *Dependencies* tab.

Reviewer: ```suggestion
. In {CENTRAL}, enter your project and click the *Settings* -> *Dependencies* tab.
```
[IBM Style Guide](https://www.ibm.com/docs/en/ibm-style?topic=elements-ui#tab) says "click" for tabs.

Author: ack

. Optional: Add any dependencies that your transformer classes require.

> . Select the *Assets* tab.

Reviewer: ```suggestion
. Click the *Assets* tab.
```
[IBM Style Guide](https://www.ibm.com/docs/en/ibm-style?topic=elements-ui#tab) says "click" for tabs.

Author: ack
